src/backend/distributed/replication/multi_logical_replication.c

```diff
@@ -2049,10 +2058,86 @@ WaitForMiliseconds(long timeout)
  * replication.
  */
 static XLogRecPtr
-GetSubscriptionPosition(MultiConnection *connection, Bitmapset *tableOwnerIds)
+GetSubscriptionPosition(MultiConnection *connection, Bitmapset *tableOwnerIds,
+						char *operationPrefix)
 {
-	char *subscriptionValueList = ShardMoveSubscriptionNamesValueList(tableOwnerIds);
+	char *subscriptionValueList = ShardSubscriptionNamesValueList(tableOwnerIds,
+																  operationPrefix);
 	return GetRemoteLSN(connection, psprintf(
 							"SELECT min(latest_end_lsn) FROM pg_stat_subscription "
 							"WHERE subname IN %s", subscriptionValueList));
 }
+
+
+/*
+ * CreateShardSplitSubscription creates the subscriptions used for shard split
+ * over the given connection. The subscription is created with 'copy_data'
+ * set to false and with the given replication slot name.
+ */
+void
+CreateShardSplitSubscription(MultiConnection *connection, char *sourceNodeName,
+							 int sourceNodePort, char *userName, char *databaseName,
+							 char *publicationName, char *slotName,
+							 Oid ownerId)
+{
+	StringInfo createSubscriptionCommand = makeStringInfo();
+	StringInfo conninfo = makeStringInfo();
+
+	/*
+	 * The CREATE USER command should not propagate, so we temporarily
+	 * disable DDL propagation.
+	 */
+	SendCommandListToWorkerOutsideTransaction(
+		connection->hostname, connection->port, connection->user,
+		list_make2(
+			"SET LOCAL citus.enable_ddl_propagation TO OFF;",
+			psprintf(
+				"CREATE USER %s SUPERUSER IN ROLE %s",
+				ShardSubscriptionRole(ownerId, SHARD_SPLIT_SUBSCRIPTION_ROLE_PREFIX),
+				quote_identifier(GetUserNameFromId(ownerId, false))
+				)));
+
+	appendStringInfo(conninfo, "host='%s' port=%d user='%s' dbname='%s' "
+							   "connect_timeout=20",
+					 escape_param_str(sourceNodeName), sourceNodePort,
+					 escape_param_str(userName), escape_param_str(databaseName));
+
+	appendStringInfo(createSubscriptionCommand,
+					 "CREATE SUBSCRIPTION %s CONNECTION %s PUBLICATION %s "
+					 "WITH (citus_use_authinfo=true, enabled=false, create_slot=false, copy_data=false, slot_name='%s')",
+					 quote_identifier(ShardSubscriptionName(ownerId,
+															SHARD_SPLIT_SUBSCRIPTION_PREFIX)),
+					 quote_literal_cstr(conninfo->data),
+					 quote_identifier(publicationName),
+					 slotName);
```

Comment:
super minor: could be nice to add escaping to be defensive

Reply:
Done